renderdoc/driver/gl/gl_driver.cpp

```diff
@@ -739,6 +742,50 @@ std::string WrappedOpenGL::GetChunkName(uint32_t idx)
   return ToStr((GLChunk)idx);
 }
 
+ResourceId WrappedOpenGL::GetContextResourceID()
+{
+  uint64_t currentId = Threading::GetCurrentID();
+  if(m_ActiveContextResourceIDs.find(currentId) == m_ActiveContextResourceIDs.end())
+  {
+    m_ActiveContextResourceIDs[currentId] =
+        GetResourceManager()->RegisterResource(GLResource(NULL, eResSpecial, eSpecialResContext));
+  }
+  return m_ActiveContextResourceIDs[currentId];
+}
+
+GLResourceRecord *WrappedOpenGL::GetContextRecord()
+{
+  ResourceId currentResourceId = GetContextResourceID();
+  if(!GetResourceManager()->HasResourceRecord(currentResourceId))
+  {
+    GLResourceRecord *contextRecord = GetResourceManager()->AddResourceRecord(currentResourceId);
+    contextRecord->DataInSerialiser = false;
+    contextRecord->Length = 0;
+    contextRecord->SpecialResource = true;
+  }
+  return GetResourceManager()->GetResourceRecord(currentResourceId);
```

Comment:
This implementation is too expensive to be called high-frequency on every function call. It also has a subtle bug where the record is associated with a thread, and not a context - we want to store chunks and whitelist/blacklist by context, not by thread, otherwise if contexts are moved between threads it would have unintended consequences.

Instead I'd recommend an implementation closer to what we talked about originally:

1. Re-use the `m_ContextData` map instead of creating a parallel map `m_ActiveContextResourceIDs` that will be 1:1 (an entry per context).
2. In `WrappedOpenGL::ActivateContext` which is much lower frequency, change the "update thread-local context pair" section of code, so that instead of storing only a `ContextPair` in the current context TLS variable, store a struct with a `ContextPair` and a `GLResourceRecord *`, and do this lazy-initialisation here.
3. Instead of using the resource manager to track whether the record is present or not, just look up in `m_ContextData` for the record - the creation can be the same but it can be stored as a member in `m_ContextData`.
3. Then `GetContextRecord()` can have a very similar implementation to `GetCtx()`, but just return the appropriate value out of the TLS variable. This should be about as fast as we can get it.
4. Probably a good idea would be to have a backup, if the TLS variable isn't initialised, to still perform this lazy creation of the record at that time. This should only happen if the user doesn't have a context active at all while calling GL commands, but knowing GL not only do a bunch of people out there do that by accident, but a handful rely on some strange quirk where it works.

That way the more expensive tracking only happens at the frequency where it might change - when a new context is bound and potentially seen.